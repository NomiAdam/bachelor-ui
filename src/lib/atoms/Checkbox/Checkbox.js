import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Heading from '../Heading/index';
import {Grid, GridCol} from '../Grid/index';
import {basicTheme, lightTheme} from '../../constants/theme';
import { LABEL_POSITION, equalsLeft, equalsRight } from './constants/position';

const StyledContainer = styled.label`
  display: block;
  position: relative;
  padding-left: 35px;
  margin-bottom: 25px;
  cursor: pointer;
  font-size: 22px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
`;

const StyledInput = styled.input`
  position: absolute;
  opacity: 0;
  cursor: pointer;
`;

const StyledSpan = styled.span`
  position: absolute;
  top: 0;
  left: 0;
  height: 25px;
  width: 25px;
  background-color: ${lightTheme.lightBlue};
  transition: .25s all ease;
  &:hover {
    background-color: ${lightTheme.whiteBlue};
  }
  ${StyledInput}:checked ~ & {
    background-color: ${basicTheme.blue};
  }
  &:after {
    content: "";
    position: absolute;
    display: none;
    left: 9px;
    top: 5px;
    width: 5px;
    height: 10px;
    border: solid ${lightTheme.white};
    border-width: 0 3px 3px 0;
    -webkit-transform: rotate(45deg);
    -ms-transform: rotate(45deg);
    transform: rotate(45deg);
  }
  ${StyledInput}:checked ~ & {
    &:after {
        display: block;
    }
  }
`;

const renderLabel = (label) => (
    <GridCol colXS={6}>
        <Heading label={label} component={'h4'} />
    </GridCol>
);

const Checkbox = ({label, checked, onChange, labelPosition}) => {
    return (
        <Grid>
            { equalsLeft(labelPosition) && renderLabel(label) }
            <GridCol colXS={6} center>
                <StyledContainer>
                    <StyledInput type="checkbox" checked={checked} onChange={onChange}/>
                    <StyledSpan/>
                </StyledContainer>
            </GridCol>
            { equalsRight(labelPosition) && renderLabel(label) }
        </Grid>
    );
};

Checkbox.propTypes = {
    /**
     * String value of label, label will reside on left side of checkbox
     */
    label: PropTypes.string,
    /**
     * Whether checkbox is checked or not
     */
    checked: PropTypes.bool,
    /**
     * Function handler of onChange event
     */
    onChange: PropTypes.func,
    /**
     * String representation of labels position
     */
    labelPosition: PropTypes.oneOf([
        LABEL_POSITION.LEFT, LABEL_POSITION.RIGHT
    ]),
};

Checkbox.defaultProps = {
    labelPosition: LABEL_POSITION.LEFT,
};

export default Checkbox;
