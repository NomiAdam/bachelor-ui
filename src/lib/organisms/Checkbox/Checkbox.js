import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { path } from 'ramda';
import Heading from '../../atoms/Heading/index';
import Container from '../Container';
import { LABEL_POSITION, equalsLeft, equalsRight } from './constants/position';
import DEFAULT_THEME, { resolveTheme, TYPES } from '../../utils/resolveTheme';

const StyledContainer = styled.label`
  display: block;
  position: relative;
  margin: ${ ({ isOnLeft }) => (isOnLeft ? '0 0 27px 15px' : '0 35px 27px 0') };
  font-size: 22px;
  user-select: none;
  cursor: ${ ({ disabled }) => (disabled ? 'not-allowed' : 'pointer') };
`;

const StyledInput = styled.input`
  position: absolute;
  opacity: 0;
`;

const StyledSpan = styled.span`
  ${ DEFAULT_THEME }
  position: absolute;
  top: 0;
  left: 0;
  height: 22px;
  width: 22px;
  transition: .25s all ease;
  &:hover {
    background-color: ${ resolveTheme(TYPES.BACKGROUND_COLOR_HOVER) };
  }
  ${ StyledInput }:checked ~ & {
  }
  &:after {
    content: "";
    position: absolute;
    display: none;
    left: 6px;
    top: 3px;
    width: 5px;
    height: 10px;
    border: solid ${ resolveTheme(TYPES.COLOR) };
    border-width: 0 3px 3px 0;
    -webkit-transform: rotate(45deg);
    -ms-transform: rotate(45deg);
    transform: rotate(45deg);
  }
  ${ StyledInput }:checked ~ & {
    &:after {
        display: block;
    }
  }
`;

const renderLabel = label => <Heading label={ label } component="h4" />;

const getTargetValue = path(['target', 'checked']);
const Checkbox = ({
    label, checked, onChange, labelPosition, disabled,
}) => (
    <Container width="100%" height="100%" justifyContent="flex-start">
        { equalsLeft(labelPosition) && renderLabel(label) }
        <StyledContainer disabled={ disabled } isOnLeft={ equalsLeft(labelPosition) }>
            <StyledInput
                type="checkbox"
                checked={ checked }
                onChange={ disabled ? undefined : e => onChange(e, getTargetValue(e)) }
                disabled={ disabled }
            />
            <StyledSpan />
        </StyledContainer>
        { equalsRight(labelPosition) && renderLabel(label) }
    </Container>
);

Checkbox.propTypes = {
    /**
   * Whether is disabled or not
   */
    disabled: PropTypes.bool,
    /**
     * String value of label, label will reside on left side of checkbox
     */
    label: PropTypes.string,
    /**
     * Whether checkbox is checked or not
     */
    checked: PropTypes.bool,
    /**
     * Function handler of onChange event
     */
    onChange: PropTypes.func,
    /**
     * String representation of labels position
     */
    labelPosition: PropTypes.oneOf([
        LABEL_POSITION.LEFT, LABEL_POSITION.RIGHT,
    ]),
};

Checkbox.defaultProps = {
    labelPosition: LABEL_POSITION.LEFT,
};

export default Checkbox;
