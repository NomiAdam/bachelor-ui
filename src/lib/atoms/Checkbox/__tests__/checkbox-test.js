/* eslint-disable no-undef */
import React from 'react';
import Checkbox from '../Checkbox';
import { shallow } from 'enzyme';

describe('Checkbox', () => {
    it('should be defined', () => {
        expect(shallow(<Checkbox />)).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<Checkbox />)).toMatchSnapshot();
    });
    it('should match the snapshot when checked', () => {
        expect(shallow(<Checkbox checked />)).toMatchSnapshot();
    });
    it('should match the snapshot with label', () => {
        expect(shallow(<Checkbox label={'Mock'} />)).toMatchSnapshot();
    });
});
