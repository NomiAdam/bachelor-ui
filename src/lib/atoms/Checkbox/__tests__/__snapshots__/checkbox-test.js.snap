// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should match the snapshot 1`] = `
<Grid__StyledGridWrapper
  gridGap="10px"
  padding="10px"
>
  <GridCol__StyledGridCol
    colXS={6}
  >
    <Heading
      component="h4"
    />
  </GridCol__StyledGridCol>
  <GridCol__StyledGridCol
    center={true}
    colXS={6}
  >
    <Checkbox__StyledContainer>
      <Checkbox__StyledInput
        type="checkbox"
      />
      <Checkbox__StyledSpan />
    </Checkbox__StyledContainer>
  </GridCol__StyledGridCol>
</Grid__StyledGridWrapper>
`;

exports[`Checkbox should match the snapshot when checked 1`] = `
<Grid__StyledGridWrapper
  gridGap="10px"
  padding="10px"
>
  <GridCol__StyledGridCol
    colXS={6}
  >
    <Heading
      component="h4"
    />
  </GridCol__StyledGridCol>
  <GridCol__StyledGridCol
    center={true}
    colXS={6}
  >
    <Checkbox__StyledContainer>
      <Checkbox__StyledInput
        checked={true}
        type="checkbox"
      />
      <Checkbox__StyledSpan />
    </Checkbox__StyledContainer>
  </GridCol__StyledGridCol>
</Grid__StyledGridWrapper>
`;

exports[`Checkbox should match the snapshot with label 1`] = `
<Grid__StyledGridWrapper
  gridGap="10px"
  padding="10px"
>
  <GridCol__StyledGridCol
    colXS={6}
  >
    <Heading
      component="h4"
      label="Mock"
    />
  </GridCol__StyledGridCol>
  <GridCol__StyledGridCol
    center={true}
    colXS={6}
  >
    <Checkbox__StyledContainer>
      <Checkbox__StyledInput
        type="checkbox"
      />
      <Checkbox__StyledSpan />
    </Checkbox__StyledContainer>
  </GridCol__StyledGridCol>
</Grid__StyledGridWrapper>
`;
